sdk/lib/conutils/pager.c

```diff
@@ -31,7 +32,188 @@
+    IN PCTCH line,
+    IN DWORD cch)
+{
+    Pager->dwFlags &= ~CON_PAGER_FLAG_DONT_OUTPUT; /* Clear the flag */
+
+    if (!Pager->PagerLine || !Pager->PagerLine(Pager, line, cch))
+        CON_STREAM_WRITE(Pager->Screen->Stream, line, cch);
+}
+
+static BOOL
+ConPagerDefaultAction(IN PCON_PAGER Pager)
+{
+    PCTCH TextBuff = Pager->TextBuff;
+    DWORD ich = Pager->ich;
+    const DWORD cch = Pager->cch;
+    DWORD iColumn = Pager->iColumn;
+    DWORD iLine = Pager->iLine;
+    DWORD lineno = Pager->lineno;
+    const DWORD ScreenColumns = Pager->ScreenColumns;
+    const DWORD ScrollRows = Pager->ScrollRows;
+    LONG nTabWidth = Pager->nTabWidth;
+    DWORD ichLast = ich;
+    UINT nCodePage;
+    BOOL IsCJK;
+    UINT nWidthOfChar = 1;
+    BOOL IsDoubleWidthCharTrailing = FALSE;
+
+    if (ich >= cch)
+        return TRUE;
+
+    nCodePage = GetConsoleOutputCP();
+    IsCJK = IsCJKCodePage(nCodePage);
+
+    /* Normalize the tab width: if negative or too large,
+     * cap it to the number of columns. */
+    if (nTabWidth < 0)
+        nTabWidth = Pager->ScreenColumns - 1;
+    else
+        nTabWidth = min(nTabWidth, Pager->ScreenColumns - 1);
+
+    if (Pager->dwFlags & CON_PAGER_FLAG_EXPAND_TABS)
+    {
+ExpandTab:
+        while ((Pager->nSpacePending > 0) && (iLine < ScrollRows))
+        {
+            ConCallPagerLine(Pager, L" ", 1);
+            --(Pager->nSpacePending);
+            ++iColumn;
+            if (iColumn % ScreenColumns == 0)
+            {
+                if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))
+                    ++iLine;
+            }
+        }
+    }
+
+    for (; ich < cch && iLine < ScrollRows; ++ich)
+    {
+        Pager->lineno = lineno;
+
+        /* TAB character */
+        if (TextBuff[ich] == TEXT('\t') &&
+            (Pager->dwFlags & CON_PAGER_FLAG_EXPAND_TABS))
+        {
+            ConCallPagerLine(Pager, &TextBuff[ichLast], ich - ichLast);
+            if (nTabWidth == 0)
+            {
+                ichLast = ich + 1;
+                continue;
+            }
+            ichLast = ++ich;
+            Pager->nSpacePending += nTabWidth - (iColumn % nTabWidth);
+            goto ExpandTab;
+        }
+
+        /* NEWLINE character */
+        if (TextBuff[ich] == TEXT('\n'))
+        {
+            ConCallPagerLine(Pager, &TextBuff[ichLast], ich - ichLast + 1);
+            ichLast = ich + 1;
+            if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))
+                ++iLine;
+            ++lineno;
+            iColumn = 0;
+            continue;
+        }
+
+        /* FORM-FEED character */
+        if (TextBuff[ich] == TEXT('\f') &&
+            (Pager->dwFlags & CON_PAGER_FLAG_EXPAND_FF))
+        {
+            /* Skip the form-feed and clear until the end of the screen */
+            ConCallPagerLine(Pager, &TextBuff[ichLast], ich - ichLast);
+            ichLast = ich + 1;
+            // FIXME: Should we handle CON_PAGER_FLAG_DONT_OUTPUT ?
```

Comment:
I would like to know your advice on that.

Reply:
We should count the logical line number, so we have to call `ConCallPagerLine` even if no output.

Reply:
Why increasing `lineno` here? This is unrelated to an actual newline in the file.

Reply:
Sorry.
https://github.com/reactos/reactos/pull/3658/files#diff-01bb03740bc237b04360f7d7054804371263fecc66b7e2b9048fea5f3e7c9a37R91-R94
```c
    if (s_dwFlags & FLAG_PLUSn) /* Skip lines */
    {
        if (Pager->lineno < s_nNextLineNo)
        {
```
Flag `FLAG_PLUSn` is watching `lineno`.